src/tasks.jl

```diff
@@ -0,0 +1,233 @@
+
+# Task copying version of fork for Trace.
+function fork(trace :: Trace, copy_vi::Function, is_ref :: Bool = false, set_retained_vns_del_by_spl!::Union{Function,Nothing} = nothing)
+    newtrace = copy(trace,copy_vi)
+    if is_ref
+        @assert set_retained_vns_del_by_spl! != nothing "[AdvancedPF] set_retained_vns_del_by_spl! is not set."
+        set_retained_vns_del_by_spl!(newtrace.vi)
+    end
+    newtrace.task.storage[:turing_trace] = newtrace
+    return newtrace
+end
+
+# PG requires keeping all randomness for the reference particle
+# Create new task and copy randomness
+function forkr(trace :: Trace, copy_vi::Function)
+
+    newtrace = Trace(trace.vi,trace.task ,trace.taskinfo,copy_vi)
+    newtrace.vi.num_produce = 0
+    return newtrace
+end
+
+current_trace() = current_task().storage[:turing_trace]
+
+const Particle = Trace
+
+"""
+Data structure for particle filters
+- effectiveSampleSize(pc :: ParticleContainer)
+- normalise!(pc::ParticleContainer)
+- consume(pc::ParticleContainer): return incremental likelihood
+"""
+mutable struct ParticleContainer{Tvi,TInfo} <: AbstractParticleContainer where {Tvi, TInfo <: AbstractTaskInfo}
+    vals::Vector{Trace{Tvi,TInfo}}
+    # A named tuple with functions to manipulate the particles vi.
+    manipulators::Dict{String,Function}
+    # logarithmic weights (Trace) or incremental log-likelihoods (ParticleContainer)
+    logWs::Vector{Float64}
+    #This corresponds to log p(x_{0:t}), therefore the log likelihood of the transitions up to this point.
+    #Especially important for ancestor sampling.
+    logpseq::Vector{Float64}
+    # log model evidence
+    logE::Float64
+    # helpful for rejuvenation steps, e.g. in SMC2
+    n_consume::Int
+
+
+end
+
+## Empty initilaizaiton
+ParticleContainer{Tvi,TInfo}() where {Tvi, TInfo <: AbstractTaskInfo} = ParticleContainer{Tvi,TInfo}(Dict{String,Function}(),0)
+ParticleContainer{Tvi,TInfo}(manipulators::Dict{String,Function}) where {Tvi, TInfo <: AbstractTaskInfo} = ParticleContainer{Tvi,TInfo}(manipulators,0)
+#Some initilaizaiton...
+function ParticleContainer{Tvi,TInfo}(manipulators::Dict{String,Function}, n::Int) where {Tvi, TInfo <: AbstractTaskInfo}
+    if !("copy" in keys(manipulators)) manipulators["copy"] = deepcopy end
+    if !("set_retained_vns_del_by_spl!" in keys(manipulators)) manipulators["set_retained_vns_del_by_spl!"] = (x) -> () end
+    ParticleContainer{Tvi,TInfo}( Vector{Trace{Tvi,TInfo}}(undef, n), manipulators, Float64[],Float64[], 0.0, 0)
+
+end
+
+
+Base.collect(pc :: ParticleContainer) = pc.vals # prev: Dict, now: Array
+Base.length(pc :: ParticleContainer)  = length(pc.vals)
+Base.similar(pc :: ParticleContainer{T}) where T = ParticleContainer{T}(0)
+# pc[i] returns the i'th particle
+Base.getindex(pc :: ParticleContainer, i :: Real) = pc.vals[i]
+
+
+
+# registers a new x-particle in the container
+function Base.push!(pc::ParticleContainer, p::Particle{Tvi,TInfo}) where {Tvi,TInfo <: AbstractTaskInfo}
```

Comment:
The static parameters are not useful.
```suggestion
function Base.push!(pc::ParticleContainer, p::Particle)
```